### PR TITLE
Adding build canceling (only works with compatible celery brokers redis and rabbitmq)

### PR DESCRIPTION
**Pull request by costrouc**
Closes https://github.com/conda-incubator/conda-store/issues/306

Build canceling exposed at `PUT /api/v1/build/<build-id>/cancel/` and added a new permissions `build::cancel` which is available by default for `admin` and `developer` roles.

### CODE REVIEW

**Comment by costrouc** [comment]
TODO: this PR requires a test:
 -  that demonstrates starting a build 
 - canceling via the api endpoint
 - seeing that the status of the build -> `FAILED`
 - and the reason in logs of the FAILURE is due to build being canceled

**Comment by trallard** [comment]
@costrouc  are you adding the test to this PR or do you need someone to take over this?

**Comment by costrouc** [comment]
@trallard I could definitely use some help and I think this would be a good experience for writing tests in conda-store.

**Comment by trallard** [comment]
Ok will assign @anirrudh  to wrap up the rest of the PR - had a look and indeed seems we are missing tests (overall) so need to prioritise that

**Comment by costrouc** [comment]
> had a look and indeed seems we are missing tests (overall) so need to prioritise that

Overview of the tests that we currently have:
  - unit tests over authentication, authorization, all build actions, celery worker tasks, traitlets, db actions `conda-store-server/tests/test_*.py`
  - integration tests (create env, update, delete etc), (and full api stability tests) `tests/test_*.py`
  - playwright browser test full flow `tests/test_playwright.py`
  - integration tests for conda-store client  `conda-store/tests`

**Comment by anirrudh** [comment]
@costrouc in that case, I will be writing tests based on the comment that you made earlier in this thread.

**Comment by anirrudh** [comment]
@costrouc Are `build_id` values still identifiers of a certain build of an environment but global (i.e. the build ids for a given env can be 1 and 3, skipping 2 as another env in the same namespace is being built but has been assigned the build_id value 2)? IIRC this is how this works.

**Comment by anirrudh** [comment]
@costrouc Are `build_id` values still identifiers of a certain build of an environment but global (i.e. the build ids for a given env can be 1 and 3, skipping 2 as another env in the same namespace is being built but has been assigned the build_id value 2)? IIRC this is how this works.

File: tests/test_api.py
Comment (costrouc) [inline tests/test_api.py:749]: This while loop will never end is some cases. Need to have this as a timeout with check interval.

File: tests/test_api.py
Comment (costrouc) [inline tests/test_api.py:767]: Should take less than 10s

**Review by costrouc — COMMENTED** [review]
I can't request changes since I started the PR :smile:. But just two tweaks in my mind. Also good to see tests are passing! So this tells me build canceling is indeed working.

File: tests/test_api.py
Comment (anirrudh) [reply tests/test_api.py]: set max to 10

Comment (anirrudh) [reply tests/test_api.py]: Tests fail when set to something low :(

File: tests/test_api.py
Comment (anirrudh) [reply tests/test_api.py]: Done, using `time`.  Good catch.

**Comment by costrouc** [comment]
LGTM. Thanks @anirrudh !

File: conda-store-server/conda_store_server/server/views/api.py
Comment (pierrotsmnrd) [inline conda-store-server/conda_store_server/server/views/api.py:725]: I see what's done here. We're cancelling all the possibly existing tasks for the given build_id. 

There's a problem with `f"build-{build_id}-conda-docker"` : The name of the task is actually `f"build-{build.id}-docker"`, based on [app.py](https://github.com/conda-incubator/conda-store/blob/67364dcd65f2b50fbea3be5be7dc4fbaadf2b7ab/conda-store-server/conda_store_server/app.py#L674)

Overall, I think we could provide a more resilient approach, where we don't have to manually maintain a list of tasks here in `api.py`, to match with `app.py`

Short term solution : fix the wrong task name above
Longer term solution : Use constants or something so the list above is generated automatically and identical to the strings used when creating the tasks, removing the need to maintain it.

**Review by pierrotsmnrd — CHANGES_REQUESTED** [review]
At least one requested change to properly cancel "docker" tasks. The rest is improvements / suggestions.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (trallard) [reply conda-store-server/conda_store_server/server/views/api.py:742]: Agree. There needs to be a clear status distinction between builds that were cancelled by the user and failed ones.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (costrouc) [reply conda-store-server/conda_store_server/server/views/api.py:725]: Totally. This makes sense. 

I'm wondering where the best place to store this is. Currently the most natural place feels like `orm.py`. 

Also related to this I am not sure if all those task names needed to be included? This is something that confuses me in celery. The "root" task name is `build-{build_id}-environment`

File: conda-store-server/conda_store_server/server/views/api.py
Comment (costrouc) [reply conda-store-server/conda_store_server/server/views/api.py:742]: I agree with this and checking celery and redis queues they both have a concept of a canceled task (revoked and canceled).

This will require a lot more file/changes. Since CANCELED is adding a new final state. Also will require an alembic migration to account for this new state.

Should we wrap this into this PR?

File: conda-store-server/conda_store_server/server/views/api.py
Comment (costrouc) [reply conda-store-server/conda_store_server/server/views/api.py:725]: @pierrotsmnrd I don't think that celery considers the children subtasks to "exist" and be cancelable/revokable. This is what needs to be checked.

Comment (costrouc) [reply conda-store-server/conda_store_server/server/views/api.py:725]: @anirrudh can you check this?

Comment (anirrudh) [reply conda-store-server/conda_store_server/server/views/api.py:725]: Yeah, I will check this in our case. I'll have some thoughts in a bit.

Comment (anirrudh) [reply conda-store-server/conda_store_server/server/views/api.py:725]: @costrouc I've experimented a bit, and this seems to be the case. Celery does not list these as part of tasks that are collected in the overall task id. I tried, and we would have to use something like `celery.chord` at task orchestration time, and then pull this information out later from the object.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/server/views/api.py:699]: Will need to be fixed after rebase. We don't use deps for db anymore.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/server/views/api.py:717]: The said docs are here (took me a bit to find): https://docs.celeryq.dev/en/latest/userguide/workers.html#worker-persistent-revokes
> Note that remote control commands must be working for revokes to work. Remote control commands are only supported by the RabbitMQ (amqp) and Redis at this point.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/server/views/api.py:730]: Move to the start of the file unless this conflicts with something.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/server/views/api.py:725]: - I agree that task ids shouldn't be hardcoded. I'll make this change.
- I'll see if it's possible to cancel subtasks using `build-{build.id}` and/or whether this task id should be canceled as well.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/server/views/api.py:742]: I was confused by the above thread a bit since the message already says "canceled". The requested change here is to pass a yet-to-be-added CANCELED (made up name) marker/param to `task_cleanup_builds`, to make use of it in `build_cleanup`, which currently just calls `set_build_failed` regardless of the reason.

Will be done in a separate PR.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:737]: This test needs an unauth version to check that auth is working, see `test_delete_build_unauth`.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:759]: This should probably be lowered to 5. No need to wait half a minute on failure.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:755]: This should also have the standard `assert r.status == schema.APIStatus.OK` to fail early if there are problems. It will fail due to building not being True, but that'd be harder to debug later.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:769]: Formatting: remove newlines here, would be easier to read near `r`.
Check the full error message in the assert.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:764]: `response.raise_for_status()` missing after this, like everywhere else.

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:781]: Note to self: when another PR adding `BuildStatus.CANCELED` is done, this should fail and will need to be updated.

**Comment by nkaretnikov** [comment]
I'm taking over this PR.

Reviewed the code (without running it) as of this commit. Looks fine overall. Will fix things outlined in the comments.

**Comment by nkaretnikov** [comment]
Note to self: rebased on a CI fix from #689, might need to rebase later in case this causes problems

File: tests/test_api.py
Comment (nkaretnikov) [inline tests/test_api.py:986]: I've seen the build status being set to COMPLETED because this was first checking too early and then too late.

**Comment by nkaretnikov** [comment]
This is ready for review.  I've tested manually as well.

Note that this commit might need to be removed: https://github.com/conda-incubator/conda-store/pull/531#issuecomment-1848796250. Waiting for someone to approve, then will merge myself.

**Comment by nkaretnikov** [comment]
@costrouc PTAL, I cannot assign you as a reviewer since it's a PR started by you.

File: conda-store-server/conda_store_server/server/views/api.py
Comment (costrouc) [inline conda-store-server/conda_store_server/server/views/api.py:988]: Could maybe use a comment to explain that we have to wait 5 seconds to ensure enough time for the task to actually be canceled.

**Comment by costrouc** [comment]
Only one "comment" about making it clear why we set a countdown of 5 seconds otherwise looks great to me! I can't approve this since I was the one that started the PR

**Review by trallard — APPROVED** [review]
Approving and once that comment is added you should be able to merge @nkaretnikov

**Pull request merged**
